### PR TITLE
feat(provider/kubernetes): runJob context images

### PR DESCRIPTION
**Pull request by ethanfrogers**
add support for using images found from context in the runJob stage. i
refactored the `Resolve from Trigger` checkbox, and included Context and
Trigger images in the same dropdown to bring it more in line with the
Server Group configuration.

spinnaker/spinnaker#1961
spinnaker/spinnaker#1960

@danielpeach can I get some early feedback on this? I had to copy a bunch of code over from the Server Group module. I feel like there may be a better way to handle that, I just don't know what it is.

### CODE REVIEW

File: app/scripts/modules/kubernetes/src/pipeline/stages/runJob/runJobStage.js [hunk 1]
```diff
@@ -76,47 +101,96 @@ module.exports = angular.module('spinnaker.kubernetes.pipeline.stage.runJobStage
         this.accounts = accounts;
       });
 
-    this.stage.cloudProvider = 'kubernetes';
-    this.stage.application = $scope.application.name;
 
-    if (!this.stage.credentials && $scope.application.defaultCredentials.kubernetes) {
-      this.stage.credentials = $scope.application.defaultCredentials.kubernetes;
+    function buildImageId(image) {
```
Comment (danielpeach) [inline app/scripts/modules/kubernetes/src/pipeline/stages/runJob/runJobStage.js]: This is just a label for the UI, right? I wouldn't want to maintain this logic in two places.

It feels like you could have an Angular filter that takes an image and returns the string, and then use that filter in the server group configuration dialogue and here.

File: app/scripts/modules/kubernetes/src/pipeline/stages/runJob/runJobStage.js [hunk 2]
```diff
@@ -62,10 +45,52 @@ module.exports = angular.module('spinnaker.kubernetes.pipeline.stage.runJobStage
...
-      this.container = buildImageDescriptor(this.stage.container.imageDescription);
+    if (!this.stage.credentials && $scope.application.defaultCredentials.kubernetes) {
+      this.stage.credentials = $scope.application.defaultCredentials.kubernetes;
     }
 
+    if(this.stage.container) {
+      this.stage.container.imageDescription.imageId = buildImageId(this.stage.container.imageDescription);
+    }
+
+    function findContextImages(current, all, visited = {}) {
```
Comment (danielpeach) [inline app/scripts/modules/kubernetes/src/pipeline/stages/runJob/runJobStage.js]: I don't know if this logic should totally be shared between the run job stage and the server group dialogue..but it does feel as if you should be able to write some function that makes this a little prettier in both places. Like

```
upstreamStagesForEach(this.stage, stages, current => {
  if (current.type === 'findImage') { ... }
  if (current.type === 'bake') { ... }
})
```

File: app/scripts/modules/kubernetes/src/pipeline/stages/runJob/runJobStage.js (hunk 1 above)
Comment (ethanfrogers) [reply app/scripts/modules/kubernetes/src/pipeline/stages/runJob/runJobStage.js]: Can you point me toward an example of something like this?

File: app/scripts/modules/kubernetes/src/pipeline/stages/runJob/runJobStage.js (hunk 2 above)
Comment (ethanfrogers) [reply app/scripts/modules/kubernetes/src/pipeline/stages/runJob/runJobStage.js]: Ah, so a shared function that handles visiting each pre-req stage and implementing the result creation as a callback?

Comment (ethanfrogers) [reply app/scripts/modules/kubernetes/src/pipeline/stages/runJob/runJobStage.js]: Also, were would I need to put something like that? Is there a file where we can put shared function like this?

**Comment by ethanfrogers** [comment]
@danielpeach I changed the `findContextImages` stuff to use `pipelineConfigService.getAllUpstreamDependencies`.

File: app/scripts/modules/kubernetes/src/pipeline/stages/runJob/runJobStage.js (hunk 1 above)
Comment (danielpeach) [reply app/scripts/modules/kubernetes/src/pipeline/stages/runJob/runJobStage.js]: Yes - I was thinking you could define a custom filter named `kubernetesImageId` (or something) and include it inside your `ui-select`:

`<span ng-bind-html="container.imageDescription | kubernetesImageId | highlight: $select.search"></span>`

You can define a filter like this: https://github.com/spinnaker/deck/blob/master/app/scripts/modules/core/src/presentation/robotToHumanFilter/robotToHuman.filter.ts

File: app/scripts/modules/kubernetes/src/pipeline/stages/runJob/runJobStage.html [hunk 3]
```diff
@@ -18,6 +18,20 @@
     </namespace-select-field>
   </stage-config-field>
 
+  <stage-config-field label="Image">
+    <ui-select name="containerSelect" ng-model="ctrl.stage.container" class="form-control input-sm" required>
+        <ui-select-match>{{ ctrl.stage.container.imageDescription | kubernetesImageId }}</ui-select-match>
+        <ui-select-choices
+          repeat="container in ctrl.containers | orderBy:'imageDescription.imageId'"
```
Comment (danielpeach) [inline app/scripts/modules/kubernetes/src/pipeline/stages/runJob/runJobStage.html]: Hmm really didn't see this `orderBy` when I made that `imageId` suggestion. What order do they end up in? The ones that come from triggers/pipeline context should be sorted to the top.

File: app/scripts/modules/kubernetes/src/presentation/imageId.filter.ts [hunk 4]
```diff
@@ -0,0 +1,37 @@
+import { module } from 'angular';
+
+export interface IDockerImage {
+    fromFindImage: boolean;
```
Comment (danielpeach) [inline app/scripts/modules/kubernetes/src/presentation/imageId.filter.ts]: Please indent with 2 spaces in this file.

File: app/scripts/modules/kubernetes/src/presentation/imageId.filter.ts [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    registry: string;
+    cluster: string;
+    pattern: string;
+    fromTrigger: boolean;
+}
+
+export function imageId(image: IDockerImage): string {
+    if (image.fromFindImage) {
+        return `${image.cluster} ${image.pattern}`;
+        } else if (image.fromBake) {
```
Comment (danielpeach) [inline app/scripts/modules/kubernetes/src/presentation/imageId.filter.ts]: Align these with the `if`

File: app/scripts/modules/kubernetes/src/pipeline/stages/runJob/runJobStage.html (hunk 3 above)
Comment (danielpeach) [reply app/scripts/modules/kubernetes/src/pipeline/stages/runJob/runJobStage.html]: Maybe that's completely determined by the `group-by` function. In that case, omit this line.

File: app/scripts/modules/kubernetes/src/pipeline/stages/runJob/runJobStage.js [hunk 6]
```diff
@@ -62,10 +49,36 @@ module.exports = angular.module('spinnaker.kubernetes.pipeline.stage.runJobStage
...
+      }
+
+      return {
+        fromContext: true,
+        fromBake: true,
+        repository: stage.ami_name,
+        organization: stage.organization,
+        stageId: stage.refId
+      };
+    });
```
Comment (danielpeach) [inline app/scripts/modules/kubernetes/src/pipeline/stages/runJob/runJobStage.js]: I think you need to filter out `undefined` from this list (in the case where you had an upstream stage whose type wasn't `findImage` or `bake`). Just tag a `.filter(image => !!image)` to the end.

File: app/scripts/modules/kubernetes/src/pipeline/stages/runJob/runJobStage.js [hunk 7]
```diff
@@ -62,10 +49,36 @@ module.exports = angular.module('spinnaker.kubernetes.pipeline.stage.runJobStage
       this.stage.dnsPolicy = 'ClusterFirst';
     }
 
-    if (this.stage.container.imageDescription.fromTrigger === true) {
```
Comment (danielpeach) [inline app/scripts/modules/kubernetes/src/pipeline/stages/runJob/runJobStage.js]: What makes this unnecessary now?

File: app/scripts/modules/kubernetes/src/pipeline/stages/runJob/runJobStage.js [hunk 8]
```diff
@@ -76,47 +89,79 @@ module.exports = angular.module('spinnaker.kubernetes.pipeline.stage.runJobStage
...
+            requests: {
+              memory: null,
+              cpu: null,
+            },
+            limits: {
+              memory: null,
+              cpu: null,
+            },
+            ports: [
+            {
```
Comment (danielpeach) [inline app/scripts/modules/kubernetes/src/pipeline/stages/runJob/runJobStage.js]: Weird formatting here (probably taken from something I did). I would move the opening and closing brackets and braces together:
```
ports: [{
  ...
}],
```

File: app/scripts/modules/kubernetes/src/pipeline/stages/runJob/runJobStage.js (hunk 7 above)
Comment (ethanfrogers) [reply app/scripts/modules/kubernetes/src/pipeline/stages/runJob/runJobStage.js]: `fromTrigger` is now set when the image is selected. It's populated as part of the `searchImages` function.

File: app/scripts/modules/kubernetes/src/pipeline/stages/runJob/runJobStage.js (hunk 6 above)
Comment (ethanfrogers) [reply app/scripts/modules/kubernetes/src/pipeline/stages/runJob/runJobStage.js]: 👍

File: app/scripts/modules/kubernetes/src/pipeline/stages/runJob/runJobStage.html (hunk 3 above)
Comment (ethanfrogers) [reply app/scripts/modules/kubernetes/src/pipeline/stages/runJob/runJobStage.html]: Yeah, they end up in a weird order. There's no sorting logic in the `group-by` function either so I need to figure out another solution.

File: app/scripts/modules/kubernetes/src/pipeline/stages/runJob/runJobStage.js [hunk 9]
```diff
@@ -76,47 +89,87 @@ module.exports = angular.module('spinnaker.kubernetes.pipeline.stage.runJobStage
...
+              hostIp: null,
+            }],
+            livenessProbe: null,
+            readinessProbe: null,
+            envVars: [],
+            command: [],
+            args: [],
+            volumeMounts: [],
+          };
+        }).sort((a,b) => {
```
Comment (ethanfrogers) [inline app/scripts/modules/kubernetes/src/pipeline/stages/runJob/runJobStage.js]: @danielpeach we can discuss this, but this is what I can up with to get around the sorting without the imageId property on each container.

File: app/scripts/modules/kubernetes/src/presentation/imageId.filter.ts [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+export function imageId(image: IDockerImage): string {
+  if (image.fromFindImage) {
+    return `${image.cluster} ${image.pattern}`;
+  } else if (image.fromBake) {
+    return `${image.repository} (Baked during execution)`;
+  } else if (image.fromTrigger && !image.tag) {
+    return `${image.registry}/${image.repository} (Tag resolved at runtime)`;
+  } else {
+    if (image.registry) {
+        return `${image.registry}/${image.repository}:${image.tag}`;
```
Comment (danielpeach) [inline app/scripts/modules/kubernetes/src/presentation/imageId.filter.ts]: Back two spaces here and below.

**Comment by ethanfrogers** [comment]
Before:
image

image


After:
image

**Pull request merged**